src/wallet/wallet.cpp

```diff
@@ -2742,215 +2740,164 @@ bool CWallet::CreateTransactionInternal(
+            if (change_spend_size == -1) {
+                coin_selection_params.change_spend_size = DUMMY_NESTED_P2WPKH_INPUT_SIZE;
+            } else {
+                coin_selection_params.change_spend_size = (size_t)change_spend_size;
+            }
 
-                    // Calculate the fees for things that aren't inputs
-                    CAmount not_input_fees = coin_selection_params.effective_fee.GetFee(coin_selection_params.tx_noinputs_size);
+            // Calculate the fees for things that aren't inputs
+            CAmount not_input_fees = coin_selection_params.effective_fee.GetFee(coin_selection_params.tx_noinputs_size);
 
-                    if (!SelectCoins(vAvailableCoins, nValueToSelect + not_input_fees, setCoins, nValueIn, coin_control, coin_selection_params))
-                    {
-                        error = _("Insufficient funds");
-                        return false;
-                    }
-                }
+            if (!SelectCoins(vAvailableCoins, nValue + not_input_fees, setCoins, selected_value, coin_control, coin_selection_params))
+            {
+                error = _("Insufficient funds");
+                return false;
+            }
 
-                const CAmount nChange = nValueIn - nValueToSelect;
-                if (nChange > 0)
-                {
-                    // Fill a vout to ourself
-                    CTxOut newTxOut(nChange, scriptChange);
+            const CAmount nChange = selected_value - nValue;
+            if (nChange > 0)
+            {
+                // Fill a vout to ourself
+                CTxOut newTxOut(nChange, scriptChange);
 
-                    CAmount cost_of_change = 0;
-                    if (pick_new_inputs) {
-                        cost_of_change = GetDiscardRate(*this).GetFee(coin_selection_params.change_spend_size) + coin_selection_params.effective_fee.GetFee(coin_selection_params.change_output_size);
-                    }
+                CAmount cost_of_change = GetDiscardRate(*this).GetFee(coin_selection_params.change_spend_size) + coin_selection_params.effective_fee.GetFee(coin_selection_params.change_output_size);
 
-                    // Never create dust outputs; if we would, just
-                    // add the dust to the fee.
-                    // When nChange is less than the cost of the change output,
-                    // send it to fees (this means BnB was used)
-                    if (IsDust(newTxOut, discard_rate) || nChange <= cost_of_change)
+                // Never create dust outputs; if we would, just
+                // add the dust to the fee.
+                // When nChange is less than the cost of the change output,
+                // send it to fees (this means BnB was used)
+                if (IsDust(newTxOut, discard_rate) || nChange <= cost_of_change)
+                {
+                    nChangePosInOut = -1;
+                    nFeeRet = nChange;
```

Comment:
this change from `+=` to `=` looks correct, just kind of sneaky. Throw in an assert that it's 0 before setting?

Reply:
Done